src/physics/systems.rs

```diff
@@ -95,39 +100,61 @@ pub fn step_world_system(
 pub fn sync_transform_system(
+    sim_to_render_time: Res<SimulationToRenderTime>,
     bodies: ResMut<RigidBodySet>,
     configuration: Res<RapierConfiguration>,
     rigid_body: &RigidBodyHandleComponent,
     mut transform: Mut<Transform>,
 ) {
+    let dt = sim_to_render_time.diff;
     if let Some(rb) = bodies.get(rigid_body.handle()) {
-        let pos = rb.position;
-
         #[cfg(feature = "dim2")]
         {
-            let rot = na::UnitQuaternion::new(na::Vector3::z() * pos.rotation.angle());
+            // Predict position at render time
+            let pos_0 = Vec2::new(
+                rb.position.translation.vector.x,
+                rb.position.translation.vector.y,
+            );
+            let vel_0 = Vec2::new(rb.linvel.x, rb.linvel.y);
+            let pos_t = pos_0 + vel_0 * dt;
+
+            // Predict rotation at render time
+            let rot_0 = Quat::from_axis_angle(Vec3::unit_z(), rb.position.rotation.angle());
+            let drot = Quat::from_rotation_z(rb.angvel.z * dt);
+            let rot_t = drot * rot_0;
+
             // Do not touch the 'z' part of the translation, used in Bevy for 2d layering
-            *transform.translation_mut().x_mut() = pos.translation.vector.x * configuration.scale;
-            *transform.translation_mut().y_mut() = pos.translation.vector.y * configuration.scale;
-            transform.set_rotation(Quat::from_xyzw(rot.i, rot.j, rot.k, rot.w));
+            *transform.translation_mut().x_mut() = pos_t.x() * configuration.scale;
+            *transform.translation_mut().y_mut() = pos_t.y() * configuration.scale;
+            transform.set_rotation(rot_t);
         }
 
         #[cfg(feature = "dim3")]
         {
-            transform.set_translation(
-                Vec3::new(
-                    pos.translation.vector.x,
-                    pos.translation.vector.y,
-                    pos.translation.vector.z,
-                ) * configuration.scale,
+            // Predict position at render time
+            let pos_0 = Vec3::new(
+                rb.position.translation.vector.x,
+                rb.position.translation.vector.y,
+                rb.position.translation.vector.z,
             );
+            let vel_0 = Vec3::new(rb.linvel.x, rb.linvel.y, rb.linvel.z);
+            let pos_t = pos_0 + vel_0 * dt;
+
+            transform.set_translation(pos_t * configuration.scale);
+
+            // Predict rotation at render time
+            if let Some(axis_na) = rb.position.rotation.axis() {
+                let axis = Vec3::new(axis_na.x, axis_na.y, axis_na.z);
+                let rot_0 = Quat::from_axis_angle(axis, rb.position.rotation.angle());
+                let drot = Quat::from_rotation_ypr(
+                    rb.angvel.y * dt,
+                    rb.angvel.x * dt,
+                    rb.angvel.z * dt,
+                );
+                let rot_t = drot * rot_0;
```

Comment:
Is this correct and will it work in all cases? Is there a simpler way with quaternions? I was thinking that maybe I can just multiply the `w` value by dt, but I'm not used to quaternions yet.